test/extensions/bootstrap/wasm/wasm_test.cc

```diff
@@ -130,61 +135,44 @@ TEST_P(WasmTest, BadSignature) {
   const auto code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(
       "{{ test_rundir }}/test/extensions/bootstrap/wasm/test_data/bad_signature_cpp.wasm"));
   EXPECT_FALSE(code.empty());
-  EXPECT_THROW_WITH_MESSAGE(wasm_->initialize(code, false), Extensions::Common::Wasm::WasmException,
-                            "Bad function signature for: proxy_on_configure");
+  EXPECT_FALSE(wasm_->initialize(code, false));
+  EXPECT_TRUE(wasm_->isFailed());
 }
 
 TEST_P(WasmTest, Segv) {
   createWasm();
   const auto code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(
       "{{ test_rundir }}/test/extensions/bootstrap/wasm/test_data/segv_cpp.wasm"));
   EXPECT_FALSE(code.empty());
-  auto context = std::make_unique<TestContext>(wasm_.get());
-  EXPECT_CALL(*context, log_(spdlog::level::err, Eq("before badptr")));
   EXPECT_TRUE(wasm_->initialize(code, false));
-
-  if (GetParam() == "v8") {
-    EXPECT_THROW_WITH_MESSAGE(
-        wasm_->startForTesting(std::move(context), plugin_),
-        Extensions::Common::Wasm::WasmException,
-        "Function: proxy_on_vm_start failed: Uncaught RuntimeError: unreachable");
-  } else if (GetParam() == "wavm") {
-    EXPECT_THROW_WITH_REGEX(wasm_->startForTesting(std::move(context), plugin_),
-                            Extensions::Common::Wasm::WasmException,
-                            "Function: proxy_on_vm_start failed: wavm.reachedUnreachable.*");
-  } else {
-    ASSERT_FALSE(true); // Neither of the above was matched.
-  }
+  auto context = static_cast<TestContext*>(wasm_->start(plugin_));
+  EXPECT_CALL(*context, log_(spdlog::level::err, Eq("before badptr")));
+  EXPECT_FALSE(wasm_->configure(context, plugin_));
+  wasm_->isFailed();
 }
 
 TEST_P(WasmTest, DivByZero) {
   createWasm();
   const auto code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(
       "{{ test_rundir }}/test/extensions/bootstrap/wasm/test_data/segv_cpp.wasm"));
   EXPECT_FALSE(code.empty());
-  auto context = std::make_unique<TestContext>(wasm_.get());
-  EXPECT_CALL(*context, log_(spdlog::level::err, Eq("before div by zero")));
   EXPECT_TRUE(wasm_->initialize(code, false));
-  wasm_->setContext(context.get());
-
+  auto context = static_cast<TestContext*>(wasm_->start(plugin_));
+  EXPECT_CALL(*context, log_(spdlog::level::err, Eq("before div by zero")));
   if (GetParam() == "v8") {
-    EXPECT_THROW_WITH_MESSAGE(
-        context->onLog(), Extensions::Common::Wasm::WasmException,
-        "Function: proxy_on_log failed: Uncaught RuntimeError: divide by zero");
-  } else if (GetParam() == "wavm") {
-    EXPECT_THROW_WITH_REGEX(context->onLog(), Extensions::Common::Wasm::WasmException,
-                            "Function: proxy_on_log failed: wavm.integerDivideByZeroOrOverflow.*");
-  } else {
-    ASSERT_FALSE(true); // Neither of the above was matched.
+    // We are not throwing out of V8 and V8 does not unwind the stack, but continues execution.
```

Comment:
Where did that change in behavior come from? This test was catching it before (see: `Uncaught RuntimeError: divide by zero` message generated by V8).

Reply:
When failures occur we no longer throw an exception because those exceptions where not caught and resulted in crashes.  Now we just set a failed flag (as V8 does).   We could throw in the special case of divide by zero and wrap everything in a try block, but I didn't go that route. As you are aware there is resistance to using exceptions in some quarters and it wasn't clear that using them in this one case was important enough to justify it.

Reply:
Sorry, let me rephrase. I don't care about exception vs failure flag.

What I don't like here is that the execution continues, which results in undefined behavior.

Unless I'm missing something, V8 is still going to trap on div by zero, instead of returning the result, and previously (i.e. before https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/29) we'd `abort()` in that case, but now we're only setting the failure flag. Shouldn't that also stop further execution?

Reply:
If I call abort in the handler then it will abort Envoy itself.  I don't know how to tell V8 to stop execution and return an error, but perhaps you do ?  I had assumed that if I just return from the trap handler then it would abort execution, but it is continuing.

Reply:
Sorry for the noise.  Something is wrong with github.  The above code is not what is actually in the branch at commit a2040d57bbb9c35ca57dfc5c5021cbfa3ea4646b.   Please pull my branch and take a look at this code.  It does in fact do was you are saying.  What you are seeing seems to be some intermediate state.